### ISSUE TITLE
Programatically adding items to fieldcollections is not working with Many-To-Many relationships

### ISSUE THREAD

**Issue by BlackbitNeueMedien** [pimcore/pimcore#3842]
I am adding entries to a fieldcollection via the PHP API. The fieldcollection contains a Many-to-many object relationship. When I add items to that field, they are not properly saved. New entries show up in the fieldcollection, but they are all empty. I am using this code:

```
$object = \Pimcore\Model\DataObject\Concrete::getById(10);
$relatedObject = \Pimcore\Model\DataObject\Concrete::getById(123);

/** @var \Pimcore\Model\DataObject\Fieldcollection $fc */
$fc = $object->getFieldcollection();

$data = new \Pimcore\Model\DataObject\Fieldcollection\Data\SomeFieldcollectionData();
$data->setRelation([$relatedObject]);

$fc->add($data);

$object->setFieldcollection($fc);

$object->save();
```

After some debugging, I found that the fieldcollection was not considered to be dirty. Since I'm using the `add` method, I figured I had to add a call to `$this->markFieldDirty('_self', true);` to that method in https://github.com/pimcore/pimcore/blob/37fe951a34de7d4d174a3f34458b05c63dc2ecea/models/DataObject/Fieldcollection.php#L156-L159
After doing so, things got weirder. I have added some logging to the code above to see the state of the MTM relation. The code used is this:

```
$printRelations = function (array $elements) {
    foreach ($elements as $element) {
        if (empty($element->getCategory())) {
            $this->io->text('Empty relation');
        } else {
            $this->io->text(count($element->getCategory()) . ' entries, first one is: ' . $element->getCategory()[0]->getFullPath());
        }
    }
};
```

It's called after initially loading the object, right after saving the object, and lastly after reloading the object. This is the result.

```
Initial state
----------

 /some/other/object/that/was/there/before

After save
----------

 /some/other/object/that/was/there/before
 /my/new/object/i/want/to/add

Object loaded again from DB
---------------------------

 Empty relation
 /my/new/object/i/want/to/add
```

So, initially the fieldcollection had already one entry. The MTM relation of that entry already has some data. After saving, it looks good at first. The collection now has two entries, the second entry is the one I just added. But after reloading the object, the MTM relation of the first entry is empty. Repeating this import will eventually leads to only empty MTM relations except for the last one:

```
Object loaded again from DB
---------------------------

 Empty relation
 Empty relation
 Empty relation
 /my/new/object/i/want/to/add

```

Now, this left me really confused. Any ideas?

**Comment by dkidd** [comment]
I've encountered this problem as well. I've found that in order to save object relationships I have to mark the FieldCollection object as dirty (i.e. `$fieldCollection->markFieldDirty('_self', true)') AND mark the field referencing the related object in the FieldCollection item dirty as well. 

Here's a code I've been using to work around the problem for now. My apologies to everyone for not just submitting a bug report—it's been a crazy month.

```
public function updateCollectionItem(DataObject\Fieldcollection\Data\AbstractData $item, DataObject\AbstractObject $object)
    {
        $getMethod = 'get' . ucfirst($item->getFieldname());
        $setMethod = 'set' . ucfirst($item->getFieldname());
        $fieldCollection = $object->$getMethod() ?: new Fieldcollection();
        if( $item->getIndex() !== null ){
            $fieldCollection->remove($item->getIndex());
        }
        $fieldCollection->add( $item );

        // If we don't mark the fieldcollection object and all its constituent items as dirty,
        // related object will be lost.
        $fieldCollection->markFieldDirty('_self', true);
        foreach( $fieldCollection->getItems() as $item ){
            $item->markFieldDirty($item->getFieldname(), true);
        }

        $object->$setMethod($fieldCollection);
    }
```

If I understand things correctly, `Pimcore\Model\DataObject\Fieldcollection\Dao::delete()` (which is called by `save`) deletes all of a FieldCollection's object relationships if the FieldCollection is dirty. See line 237 of `pimcore/models/DataObject/Fieldcollection/Dao.php`. The result is that newly added FieldCollection items get saved properly, but old ones aren't saved (unless they're explicitly marked dirty).

**Comment by brusch** [comment]
Fixed by #3845
